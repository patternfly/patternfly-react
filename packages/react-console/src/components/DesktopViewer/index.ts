export * from './DesktopViewer';
export * from './MoreInformationDefaultContent';
export * from './MoreInformationInstallVariant';
