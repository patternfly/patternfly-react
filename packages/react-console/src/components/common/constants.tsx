const NONE_TYPE = '_none_';
const SERIAL_CONSOLE_TYPE = 'SerialConsole';
const SPICE_CONSOLE_TYPE = 'SpiceConsole';
const VNC_CONSOLE_TYPE = 'VncConsole';
const RDP_CONSOLE_TYPE = 'RdpConsole';
const DESKTOP_VIEWER_CONSOLE_TYPE = 'DesktopViewer';

const CONNECTING = 'connecting';
const CONNECTED = 'connected';
const DISCONNECTED = 'disconnected';
const LOADING = 'loading';

const DEFAULT_VV_FILENAME = 'console.vv';
const DEFAULT_VV_MIMETYPE = 'application/x-virt-viewer';
const DEFAULT_RDP_FILENAME = 'console.rdp';
const DEFAULT_RDP_MIMETYPE = 'application/rdp';
const DEFAULT_RDP_PORT = 3389;

export const constants = {
  NONE_TYPE,
  SERIAL_CONSOLE_TYPE,
  SPICE_CONSOLE_TYPE,
  VNC_CONSOLE_TYPE,
  RDP_CONSOLE_TYPE,
  DESKTOP_VIEWER_CONSOLE_TYPE,

  CONNECTING,
  CONNECTED,
  DISCONNECTED,
  LOADING,

  DEFAULT_VV_FILENAME,
  DEFAULT_VV_MIMETYPE,
  DEFAULT_RDP_FILENAME,
  DEFAULT_RDP_MIMETYPE,
  DEFAULT_RDP_PORT
};
