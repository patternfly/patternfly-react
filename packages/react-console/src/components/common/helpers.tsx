import React from 'react';

export const child = (children: React.ReactNode, index: number) => {
  const childrenArray = React.Children.toArray(children);
  return (childrenArray && childrenArray.length > index && childrenArray[index]) || null;
};
