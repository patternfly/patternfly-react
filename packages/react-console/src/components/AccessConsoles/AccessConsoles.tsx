import React from 'react';
import { css } from '@patternfly/react-styles';
import { Select, SelectOption, SelectOptionObject, SelectVariant } from '@patternfly/react-core';

import { constants } from '../common/constants';

import styles from '@patternfly/react-styles/css/components/Consoles/AccessConsoles';
import '@patternfly/react-styles/css/components/Consoles/AccessConsoles.css';

const { NONE_TYPE, SERIAL_CONSOLE_TYPE, VNC_CONSOLE_TYPE, DESKTOP_VIEWER_CONSOLE_TYPE } = constants;

const getChildTypeName = (child: any) =>
  child && child.props && child.props.type ? child.props.type : (child && child.type && child.type.displayName) || null;

const isChildOfType = (child: any, type: string) => {
  if (child && child.props && child.props.type) {
    return child.props.type === type;
  } else if (child && child.type) {
    return child.type.displayName === type;
  }
  return false;
};

export interface AccessConsolesProps {
  /**
   * Child element can be either
   *   - <SerialConsole>, <VncConsole> or <DesktopViewer>
   *   - or has a property "type" of value either SERIAL_CONSOLE_TYPE or VNC_CONSOLE_TYPE (useful when wrapping (composing) basic console components
   */
  children?: React.ReactElement[] | React.ReactNode;
  /** Placeholder text for the console selection */
  textSelectConsoleType?: string;
  /** The value for the Serial Console option. This can be overriden by the type property of the child component */
  textSerialConsole?: string;
  /** The value for the VNC Console option. This can be overriden by the type property of the child component */
  textVncConsole?: string;
  /** The value for the Desktop Viewer Console option. This can be overriden by the type property of the child component */
  textDesktopViewerConsole?: string;
  /** Initial selection of the Select */
  preselectedType?: string; // NONE_TYPE | SERIAL_CONSOLE_TYPE | VNC_CONSOLE_TYPE | DESKTOP_VIEWER_CONSOLE_TYPE;
}

export const AccessConsoles: React.FunctionComponent<AccessConsolesProps> = ({
  children,
  textSelectConsoleType = 'Select console type',
  textSerialConsole = 'Serial console',
  textVncConsole = 'VNC console',
  textDesktopViewerConsole = 'Desktop viewer',
  preselectedType = null
}) => {
  const typeMap = {
    [SERIAL_CONSOLE_TYPE]: textSerialConsole,
    [VNC_CONSOLE_TYPE]: textVncConsole,
    [DESKTOP_VIEWER_CONSOLE_TYPE]: textDesktopViewerConsole
  };
  const [type, setType] = React.useState(
    preselectedType !== NONE_TYPE
      ? ({ value: preselectedType, toString: () => typeMap[preselectedType] } as SelectOptionObject)
      : null
  );
  const [isOpen, setIsOpen] = React.useState(false);

  const getConsoleForType = (type: any) =>
    React.Children.map(children as React.ReactElement[], (child: any) => {
      if (getChildTypeName(child) === type.value) {
        return <React.Fragment key={getChildTypeName(child)}>{child}</React.Fragment>;
      } else {
        return null;
      }
    });

  const onToggle = (isOpen: boolean) => {
    setIsOpen(isOpen);
  };

  const selectOptions: any[] = [];

  React.Children.toArray(children as React.ReactElement[]).forEach((child: any) => {
    if (isChildOfType(child, VNC_CONSOLE_TYPE)) {
      selectOptions.push(
        <SelectOption
          key={VNC_CONSOLE_TYPE}
          id={VNC_CONSOLE_TYPE}
          value={{ value: VNC_CONSOLE_TYPE, toString: () => textVncConsole } as SelectOptionObject}
        />
      );
    } else if (isChildOfType(child, SERIAL_CONSOLE_TYPE)) {
      selectOptions.push(
        <SelectOption
          key={SERIAL_CONSOLE_TYPE}
          id={SERIAL_CONSOLE_TYPE}
          value={{ value: SERIAL_CONSOLE_TYPE, toString: () => textSerialConsole } as SelectOptionObject}
        />
      );
    } else if (isChildOfType(child, DESKTOP_VIEWER_CONSOLE_TYPE)) {
      selectOptions.push(
        <SelectOption
          key={DESKTOP_VIEWER_CONSOLE_TYPE}
          id={DESKTOP_VIEWER_CONSOLE_TYPE}
          value={{ value: DESKTOP_VIEWER_CONSOLE_TYPE, toString: () => textDesktopViewerConsole } as SelectOptionObject}
        />
      );
    } else {
      const typeText = getChildTypeName(child);
      selectOptions.push(
        <SelectOption
          key={typeText}
          id={typeText}
          value={{ value: typeText, toString: () => typeText } as SelectOptionObject}
        />
      );
    }
  });
  return (
    <div className={css(styles.console)}>
      {React.Children.toArray(children).length > 1 && (
        <div className={css(styles.consoleActions)}>
          <Select
            aria-label={textSelectConsoleType}
            placeholderText={textSelectConsoleType}
            toggleId="pf-c-console__type-selector"
            variant={SelectVariant.single}
            onSelect={(_, selection, __) => {
              setType(selection as SelectOptionObject);
              setIsOpen(false);
            }}
            selections={type}
            isOpen={isOpen}
            onToggle={onToggle}
          >
            {selectOptions}
          </Select>
        </div>
      )}
      {type && getConsoleForType(type)}
    </div>
  );
};
AccessConsoles.displayName = 'AccessConsoles';
