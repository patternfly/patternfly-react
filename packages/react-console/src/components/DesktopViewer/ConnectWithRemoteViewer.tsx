import React from 'react';
import { Button, ExpandableSection } from '@patternfly/react-core';

import { constants } from '../common/constants';

import { MoreInformationDefaultContent } from './MoreInformationDefaultContent';
import { MoreInformationDefaultRDPContent } from './MoreInformationDefaultRDPContent';

import {
  generateDescriptorFile,
  downloadFile,
  onGenerateFunctionType,
  onDownloadFunctionType
} from './consoleDescriptorGenerator';
import { ConsoleDetailPropType } from './ConsoleDetailPropType';

const {
  VNC_CONSOLE_TYPE,
  SPICE_CONSOLE_TYPE,
  RDP_CONSOLE_TYPE,
  DEFAULT_VV_FILENAME,
  DEFAULT_VV_MIMETYPE,
  DEFAULT_RDP_FILENAME,
  DEFAULT_RDP_MIMETYPE
} = constants;

export interface ConnectWithRemoteViewerProps extends React.HTMLProps<HTMLDivElement> {
  /** Custom content of more-info section  */
  children?: React.ReactNode;

  /** Connection details for Spice */
  spice?: ConsoleDetailPropType;
  /** Connection details for VNC */
  vnc?: ConsoleDetailPropType;
  /** Connection details for RDP */
  rdp?: ConsoleDetailPropType;

  /** Callback function. Generate content of .vv file.
   * Parameters: ({ _console, type }) => ({
   *     content,  // required string value
   *     mimeType, // optional, default application/x-virt-viewer
   *     fileName  // optional, default: console.vv
   *   })
   */
  onGenerate?: onGenerateFunctionType;
  /** Callback function. Perform file download.
   * Default implementation is usually good enough, but i.e. in case of environment with tight
   * content security policy set, this might be required.
   *
   * Examples for alternative file-download implementations:
   *   - use of iframe
   *   - use of http-server
   *
   * Parameters: (fileName, content, mimeType) => {}
   */
  onDownload?: onDownloadFunctionType;

  textConnectWithRemoteViewer?: string;
  textMoreInfo?: string;
  textMoreInfoContent?: string | React.ReactNode;
  textConnectWithRDP?: string;
  textMoreRDPInfo?: string;
  textMoreRDPInfoContent?: string | React.ReactNode;
}

export const ConnectWithRemoteViewer: React.FunctionComponent<ConnectWithRemoteViewerProps> = ({
  onGenerate = generateDescriptorFile,
  onDownload = downloadFile,
  spice = null,
  vnc = null,
  rdp = null,
  textConnectWithRemoteViewer = 'Launch Remote Viewer',
  textConnectWithRDP = 'Launch Remote Desktop',
  textMoreInfo = 'Remote Viewer Details',
  textMoreInfoContent = '',
  textMoreRDPInfo = 'Remote Desktop Details',
  textMoreRDPInfoContent = ''
}: ConnectWithRemoteViewerProps) => {
  const [isExpandedDefault, setIsExpandedDefault] = React.useState(false);
  const [isExpandedRDP, setIsExpandedRDP] = React.useState(false);

  const _console = spice || vnc; // strictly prefer spice over vnc

  const onClickVV = () => {
    const type = spice ? SPICE_CONSOLE_TYPE : VNC_CONSOLE_TYPE;
    if (_console) {
      const vv = onGenerate(_console, type);
      return onDownload(DEFAULT_VV_FILENAME, vv.content, vv.mimeType || DEFAULT_VV_MIMETYPE);
    }
  };

  const onClickRDP = () => {
    const rdpFile = onGenerate(rdp, RDP_CONSOLE_TYPE);
    return onDownload(DEFAULT_RDP_FILENAME, rdpFile.content, rdpFile.mimeType || DEFAULT_RDP_MIMETYPE);
  };

  // RDP button is rendered only if the protocol is available
  // If none of Spice or VNC is available, the .vv button is disabled (but rendered)
  return (
    <div className="pf-c-console__remote-viewer">
      <div className="pf-c-console__remote-viewer-launch">
        <Button className="pf-c-console__remote-viewer-launch-vv" onClick={onClickVV} isDisabled={!_console}>
          {textConnectWithRemoteViewer}
        </Button>
        {!!rdp && (
          <Button onClick={onClickRDP} className="pf-c-console__remote-viewer-launch-rdp">
            {textConnectWithRDP}
          </Button>
        )}
      </div>
      {!!_console && (
        <ExpandableSection
          toggleText={textMoreInfo}
          isExpanded={isExpandedDefault}
          onToggle={isExpanded => setIsExpandedDefault(isExpanded)}
        >
          <MoreInformationDefaultContent textMoreInfoContent={textMoreInfoContent} />
        </ExpandableSection>
      )}
      {!!rdp && (
        <ExpandableSection
          toggleText={textMoreRDPInfo}
          isExpanded={isExpandedRDP}
          onToggle={isExpanded => setIsExpandedRDP(isExpanded)}
        >
          <MoreInformationDefaultRDPContent textMoreRDPInfoContent={textMoreRDPInfoContent} />
        </ExpandableSection>
      )}
    </div>
  );
};
ConnectWithRemoteViewer.displayName = 'ConnectWithRemoteViewer';
