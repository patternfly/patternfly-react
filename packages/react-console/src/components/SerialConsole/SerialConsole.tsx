import React from 'react';

import { css } from '@patternfly/react-styles';
import { Button, EmptyState, EmptyStateBody, EmptyStateIcon, Spinner } from '@patternfly/react-core';

import { XTerm, XTermProps } from './XTerm';
import { SerialConsoleActions } from './SerialConsoleActions';

import { constants } from '../common/constants';

import styles from '@patternfly/react-styles/css/components/Consoles/SerialConsole';
import '@patternfly/react-styles/css/components/Consoles/xterm.css';
import '@patternfly/react-styles/css/components/Consoles/SerialConsole.css';

const { CONNECTED, DISCONNECTED, LOADING } = constants;

export interface SerialConsoleProps extends XTermProps {
  /** Initiate connection to backend. In other words, the calling components manages connection state. */
  onConnect: () => void;
  /** Close connection to backend */
  onDisconnect: () => void;
  /** Terminal produced data, like key-press */
  onData: (e: string) => void;
  /** Terminal title has been changed */
  onTitleChanged?: () => void;
  /** Connection status; a value from [''connected'; 'disconnected'; 'loading']. Default is 'loading' for a not matching value. */
  /** The number of columns to resize to */
  cols?: number;
  /** The number of rows to resize to */
  rows?: number;
  fontFamily?: string;
  fontSize?: number;
  status?: string;
  /** Text content rendered inside the Connect button */
  textConnect?: string;
  /** Text content rendered inside the Disconnect button */
  textDisconnect?: string;
  /** Text content rendered inside the Reset button */
  textReset?: string;
  /* Text content rendered inside the EmptyState for when console is disconnnected */
  textDisconnected?: string;
  /* Text content rendered inside the EmptyState for when console is loading */
  textLoading?: string;
  /** A reference object to attach to the SerialConsole. */
  innerRef?: React.RefObject<any>;
}

const SerialConsoleBase: React.FunctionComponent<SerialConsoleProps> = ({
  onConnect,
  onDisconnect,
  onTitleChanged = () => {},
  onData,
  cols,
  rows,
  fontFamily,
  fontSize,
  status = 'loading',
  textConnect = 'Connect',
  textDisconnect,
  textReset,
  textDisconnected = 'Click Connect to open serial console.',
  textLoading = 'Loading ...',
  innerRef
}) => {
  React.useEffect(() => {
    onConnect();
    return () => {
      onDisconnect();
    };
  }, []);

  const onConnectClick = () => {
    onConnect();
    focusTerminal();
  };

  const onDisconnectClick = () => {
    onDisconnect();
    focusTerminal();
  };

  const onResetClick = () => {
    onDisconnect();
    onConnect();
    focusTerminal();
  };

  const focusTerminal = () => {
    innerRef && innerRef.current && innerRef.current.focusTerminal();
  };

  let terminal;
  switch (status) {
    case CONNECTED:
      terminal = (
        <XTerm
          innerRef={innerRef}
          cols={cols}
          rows={rows}
          fontFamily={fontFamily}
          fontSize={fontSize}
          onTitleChanged={onTitleChanged}
          onData={onData}
        />
      );
      break;
    case DISCONNECTED:
      terminal = (
        <EmptyState>
          <EmptyStateBody>{textDisconnected}</EmptyStateBody>
          <Button onClick={onConnectClick}>{textConnect}</Button>
        </EmptyState>
      );
      break;
    case LOADING:
    default:
      terminal = (
        <EmptyState>
          <EmptyStateIcon variant="container" component={Spinner} />
          <EmptyStateBody>{textLoading}</EmptyStateBody>
        </EmptyState>
      );
      break;
  }

  return (
    <>
      {status !== DISCONNECTED && (
        <SerialConsoleActions
          onDisconnect={onDisconnectClick}
          onReset={onResetClick}
          textDisconnect={textDisconnect}
          textReset={textReset}
        />
      )}
      <div className={css(styles.consoleSerial)}>{terminal}</div>
    </>
  );
};
SerialConsoleBase.displayName = 'SerialConsoleBase';
export const SerialConsole = React.forwardRef((props: SerialConsoleProps, ref: React.Ref<HTMLDivElement>) => (
  <SerialConsoleBase innerRef={ref as React.MutableRefObject<any>} {...props} />
));
SerialConsole.displayName = 'SerialConsole';
