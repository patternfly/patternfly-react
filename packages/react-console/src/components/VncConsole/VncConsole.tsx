import React from 'react';

import { css } from '@patternfly/react-styles';
import { EmptyState, EmptyStateBody, EmptyStateIcon, Spinner } from '@patternfly/react-core';

import * as NovncLog from 'novnc-core/lib/util/logging';
/** Has bad types. https://github.com/larryprice/novnc-core/issues/5 */
import RFB from 'novnc-core';

import { VncActions } from './VncActions';
import constants from '../common/constants';

import styles from '@patternfly/react-styles/css/components/Consoles/VncConsole';
import '@patternfly/react-styles/css/components/Consoles/VncConsole.css';

const { CONNECTED, CONNECTING, DISCONNECTED } = constants;

export interface VncConsoleProps extends React.HTMLProps<HTMLDivElement> {
  /** Children nodes */
  children?: React.ReactNode;

  /** FQDN or IP to connect to */
  host: string;
  /** TCP Port */
  port?: string;
  /** host:port/path */
  path?: string;
  encrypt?: boolean;
  /** Is a boolean indicating if a request to resize the remote session should be sent whenever the container changes dimensions */
  resizeSession?: boolean;
  /** Is a boolean indicating if the remote session should be scaled locally so it fits its container */
  scaleViewport?: boolean;
  /** Is a boolean indicating if any events (e.g. key presses or mouse movement) should be prevented from being sent to the server */
  viewOnly?: boolean;
  /** Is a boolean indicating if the remote server should be shared or if any other connected clients should be disconnected */
  shared?: boolean;
  /** An Object specifying the credentials to provide to the server when authenticating
   * { username: '' password: '' target: ''}
   */
  credentials?: object;
  /** A DOMString specifying the ID to provide to any VNC repeater encountered */
  repeaterID?: string;
  /** log-level for noVNC */
  vncLogging?: 'error' | 'warn' | 'none' | 'debug' | 'info';
  consoleContainerId?: string;
  additionalButtons?: React.ReactNode[];

  /** Callback. VNC server disconnected. */
  onDisconnected?: (e: any) => void;
  /** Initialization of RFB failed */
  onInitFailed?: (e: any) => void;
  /** Handshake failed */
  onSecurityFailure?: (e: any) => void;

  /* Text content rendered inside the EmptyState for when console is connecting */
  textConnecting?: string | React.ReactNode;
  /* Text content rendered inside the EmptyState for when console is disconnnected */
  textDisconnected?: string;
  /** Text content rendered inside the Disconnect button */
  textDisconnect?: string;
  /** Text content rendered inside the button Send shortcut dropdown toggle */
  textSendShortcut?: string;
  /** Text content rendered inside the Ctrl-Alt-Delete dropdown entry */
  textCtrlAltDel?: string;
}

export const VncConsole: React.FunctionComponent<VncConsoleProps> = ({
  children,
  host,
  port = '80',
  path = '',
  encrypt = false,
  resizeSession = true,
  scaleViewport = false,
  viewOnly = false,
  shared = false,
  credentials,
  repeaterID = '',
  vncLogging = 'warn',
  consoleContainerId,
  additionalButtons = [] as React.ReactNode[],
  onDisconnected = () => {},
  onInitFailed,
  onSecurityFailure,
  textConnecting = 'Connecting',
  textDisconnected = 'Disconnected',
  textDisconnect = 'Disconnect',
  textSendShortcut,
  textCtrlAltDel
}) => {
  let rfb: any;
  let novncStaticComponent: React.ReactNode;
  let novncElem: HTMLDivElement;

  const [status, setStatus] = React.useState(CONNECTING);

  const addEventListeners = () => {
    rfb.addEventListener('connect', onConnected);
    rfb.addEventListener('disconnect', onDisconnected);
    rfb.addEventListener('securityfailure', onSecurityFailure);
  };

  React.useEffect(() => {
    NovncLog.init_logging(vncLogging);
    try {
      const protocol = encrypt ? 'wss' : 'ws';
      const url = `${protocol}://${host}:${port}/${path}`;

      const options = {
        repeaterID,
        shared,
        credentials
      };
      rfb = (new RFB() as any)(novncElem, url, options);
      addEventListeners();
      rfb.viewOnly = viewOnly;
      rfb.scaleViewport = scaleViewport;
      rfb.resizeSession = resizeSession;
    } catch (e) {
      onInitFailed && onInitFailed(e);
      rfb = undefined;
    }

    return () => {
      disconnect();
      removeEventListeners();
      rfb = undefined;
    };
  }, []);

  const disconnect = () => {
    if (!rfb) {
      return;
    }
    rfb.disconnect();
  };

  const onConnected = () => {
    setStatus(CONNECTED);
  };

  const _onDisconnected = (e: any) => {
    setStatus(DISCONNECTED);
    onDisconnected(e);
  };

  const _onSecurityFailure = (e: any) => {
    setStatus(DISCONNECTED);
    onSecurityFailure(e);
  };

  const onCtrlAltDel = () => {
    if (rfb) {
      rfb.sendCtrlAltDel();
    }
  };

  const removeEventListeners = () => {
    if (rfb) {
      rfb.removeEventListener('connect', onConnected);
      rfb.removeEventListener('disconnect', _onDisconnected);
      rfb.removeEventListener('securityfailure', _onSecurityFailure);
    }
  };

  let rightContent;
  let emptyState;
  switch (status) {
    case CONNECTED:
      rightContent = (
        <VncActions
          onCtrlAltDel={onCtrlAltDel}
          textSendShortcut={textSendShortcut}
          textCtrlAltDel={textCtrlAltDel}
          textDisconnect={textDisconnect}
          onDisconnect={disconnect}
          additionalButtons={additionalButtons}
        />
      );
      break;
    case DISCONNECTED:
      emptyState = (
        <EmptyState>
          <EmptyStateBody>{textDisconnected}</EmptyStateBody>
        </EmptyState>
      );
      break;
    case CONNECTING:
    default:
      emptyState = (
        <EmptyState>
          <EmptyStateIcon variant="container" component={Spinner} />
          <EmptyStateBody>{textConnecting}</EmptyStateBody>
        </EmptyState>
      );
  }

  if (!novncStaticComponent) {
    novncStaticComponent = <div id={consoleContainerId} ref={e => (novncElem = e)} />;
  }

  return (
    <div className={css(styles.consoleVnc)}>
      {children}
      <React.Fragment>
        {rightContent}
        <div>
          {emptyState}
          {novncStaticComponent}
        </div>
      </React.Fragment>
    </div>
  );
};
VncConsole.displayName = 'VncConsole';
