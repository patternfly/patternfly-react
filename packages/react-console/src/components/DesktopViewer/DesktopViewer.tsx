import React from 'react';

import { css } from '@patternfly/react-styles';
import { ManualConnection } from './ManualConnection';
import { ConnectWithRemoteViewer, ConnectWithRemoteViewerProps } from './ConnectWithRemoteViewer';
import { ConsoleDetailPropType } from './ConsoleDetailPropType';

import styles from '@patternfly/react-styles/css/components/Consoles/DesktopViewer';
import '@patternfly/react-styles/css/components/Consoles/DesktopViewer.css';

export interface DesktopViewerProps extends ConnectWithRemoteViewerProps {
  /** Custom content of more-info section  */
  children?: React.ReactNode;
  /** Connection details for Spice */
  spice?: ConsoleDetailPropType;
  /** Connection details for VNC */
  vnc?: ConsoleDetailPropType;
  /** Connection details for RDP */
  rdp?: ConsoleDetailPropType;

  textManualConnection?: string;
  textNoProtocol?: string;
  textConnectWith?: string;

  textAddress?: string;
  textSpiceAddress?: string;
  textVNCAddress?: string;
  textSpicePort?: string;
  textVNCPort?: string;
  textSpiceTlsPort?: string;
  textVNCTlsPort?: string;
  textRDPPort?: string;
  textRdpAddress?: string;

  textConnectWithRemoteViewer?: string;
  textConnectWithRDP?: string;
  /** Text that appears in the toggle */
  textMoreInfo?: string;
  /** The information content appearing above the description list for guidelines to install virt-viewer */
  textMoreInfoContent?: string | React.ReactNode;
  /** Text that appears in the toggle */
  textMoreRDPInfo?: string;
  /** The information content appearing above the description list for guidelines to install virt-viewer */
  textMoreRDPInfoContent?: string | React.ReactNode;
}

export const DesktopViewer: React.FunctionComponent<DesktopViewerProps> = ({
  children = null,
  spice = null,
  vnc = null,
  rdp = null,
  ...props
}: DesktopViewerProps) => (
  <div className={css(styles.consoleDesktopViewer)}>
    <ConnectWithRemoteViewer
      spice={spice}
      vnc={vnc}
      rdp={rdp}
      onGenerate={props.onGenerate}
      onDownload={props.onDownload}
      textConnectWithRemoteViewer={props.textConnectWithRemoteViewer}
      textConnectWithRDP={props.textConnectWithRDP}
      textMoreInfo={props.textMoreInfo}
      textMoreRDPInfo={props.textMoreRDPInfo}
      textMoreInfoContent={props.textMoreInfoContent}
      textMoreRDPInfoContent={props.textMoreRDPInfoContent}
    >
      {children}
    </ConnectWithRemoteViewer>
    <ManualConnection
      spice={spice}
      vnc={vnc}
      rdp={rdp}
      textManualConnection={props.textManualConnection}
      textNoProtocol={props.textNoProtocol}
      textConnectWith={props.textConnectWith}
      textAddress={props.textAddress}
      textSpiceAddress={props.textSpiceAddress}
      textVNCAddress={props.textVNCAddress}
      textSpicePort={props.textSpicePort}
      textVNCPort={props.textVNCPort}
      textSpiceTlsPort={props.textSpiceTlsPort}
      textVNCTlsPort={props.textVNCTlsPort}
      textRDPPort={props.textRDPPort}
      textRdpAddress={props.textRdpAddress}
    />
  </div>
);
DesktopViewer.displayName = 'DesktopViewer';
