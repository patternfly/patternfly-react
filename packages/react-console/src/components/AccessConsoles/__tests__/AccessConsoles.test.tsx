import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { AccessConsoles } from '../AccessConsoles';
import { SerialConsole } from '../../SerialConsole';
import { VncConsole } from '../../VncConsole';
import { DesktopViewer } from '../../DesktopViewer';
import { constants } from '../../common/constants';

const { SERIAL_CONSOLE_TYPE, LOADING } = constants;

const MyVncConsoleTestWrapper = () => <p>VNC console text</p>;
const SerialConsoleConnected = () => <p>Serial console text</p>;

const vnc = {
  address: 'my.host.com',
  port: 5902,
  tlsPort: '5903'
};

describe('AccessConsoles', () => {
  beforeAll(() => {
    window.HTMLCanvasElement.prototype.getContext = () => ({ canvas: {} } as any);
  });

  test('with SerialConsole as a single child', () => {
    const { asFragment } = render(
      <AccessConsoles>
        <SerialConsole onData={jest.fn()} onConnect={jest.fn()} onDisconnect={jest.fn()} status={LOADING} />
      </AccessConsoles>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('with VncConsole as a single child', () => {
    const { asFragment } = render(
      <AccessConsoles>
        <VncConsole host="foo.bar.host" textDisconnected="Disconnected state text" />
      </AccessConsoles>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('with SerialConsole and VncConsole as children', () => {
    const { asFragment } = render(
      <AccessConsoles>
        <VncConsole host="foo.bar.host" textDisconnected="Disconnected state text" />
        <SerialConsole onData={jest.fn()} onConnect={jest.fn()} onDisconnect={jest.fn()} status={LOADING} />
      </AccessConsoles>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('with wrapped SerialConsole as a child', () => {
    const { asFragment } = render(
      <AccessConsoles>
        <SerialConsoleConnected />
      </AccessConsoles>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('with preselected SerialConsole', () => {
    const { asFragment } = render(
      <AccessConsoles preselectedType={SERIAL_CONSOLE_TYPE}>
        <SerialConsole onData={jest.fn()} onConnect={jest.fn()} onDisconnect={jest.fn()} status={LOADING} />
      </AccessConsoles>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('switching SerialConsole and VncConsole', () => {
    render(
      <AccessConsoles>
        <MyVncConsoleTestWrapper />
        <SerialConsole onData={jest.fn()} onConnect={jest.fn()} onDisconnect={jest.fn()} status={LOADING} />
      </AccessConsoles>
    );

    // VNC (first option) is initially selected
    expect(screen.queryByText(/Loading/)).toBeNull();
    expect(screen.getByText('VNC console text')).toBeInTheDocument();

    // Open dropdown and select "Serial console" option
    userEvent.click(screen.getByRole('button', { name: 'Options menu' }));
    userEvent.click(screen.getByText('Serial console', { selector: 'button' }));

    // VNC content is no longer visible, and loading contents of the Serial console are visible.
    expect(screen.getByText(/Loading/)).toBeInTheDocument();
    expect(screen.queryByText('VNC console text')).toBeNull();
  });

  test('Empty', () => {
    const { asFragment } = render(<AccessConsoles />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('with DesktopViewer', () => {
    const { asFragment } = render(
      <AccessConsoles>
        <DesktopViewer vnc={vnc} />
      </AccessConsoles>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
