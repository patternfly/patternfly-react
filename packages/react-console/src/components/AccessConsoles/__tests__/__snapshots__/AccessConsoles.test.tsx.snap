// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessConsoles Empty 1`] = `
<DocumentFragment>
  <div
    class="pf-c-console"
  />
</DocumentFragment>
`;

exports[`AccessConsoles with DesktopViewer 1`] = `
<DocumentFragment>
  <div
    class="pf-c-console"
  />
</DocumentFragment>
`;

exports[`AccessConsoles with SerialConsole and VncConsole as children 1`] = `
<DocumentFragment>
  <div
    class="pf-c-console"
  >
    <div
      class="pf-c-console__actions"
    >
      <div
        class="pf-c-select"
        data-ouia-component-id="OUIA-Generated-Select-single-1"
        data-ouia-component-type="PF4/Select"
        data-ouia-safe="true"
      >
        <button
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-label="Options menu"
          aria-labelledby=" pf-c-console__type-selector"
          class="pf-c-select__toggle"
          id="pf-c-console__type-selector"
          type="button"
        >
          <div
            class="pf-c-select__toggle-wrapper"
          >
            <span
              class="pf-c-select__toggle-text"
            >
              Select console type
            </span>
          </div>
          <span
            class="pf-c-select__toggle-arrow"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AccessConsoles with SerialConsole as a single child 1`] = `
<DocumentFragment>
  <div
    class="pf-c-console"
  />
</DocumentFragment>
`;

exports[`AccessConsoles with VncConsole as a single child 1`] = `
<DocumentFragment>
  <div
    class="pf-c-console"
  />
</DocumentFragment>
`;

exports[`AccessConsoles with preselected SerialConsole 1`] = `
<DocumentFragment>
  <div
    class="pf-c-console"
  >
    <div
      class="pf-c-console__actions-serial"
    >
      <button
        aria-disabled="false"
        class="pf-c-button pf-m-secondary"
        data-ouia-component-id="OUIA-Generated-Button-secondary-1"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        type="button"
      >
        Disconnect
      </button>
      <button
        aria-disabled="false"
        class="pf-c-button pf-m-secondary"
        data-ouia-component-id="OUIA-Generated-Button-secondary-2"
        data-ouia-component-type="PF4/Button"
        data-ouia-safe="true"
        type="button"
      >
        Reset
      </button>
    </div>
    <div
      class="pf-c-console__serial"
    >
      <div
        class="pf-c-empty-state"
      >
        <div
          class="pf-c-empty-state__content"
        >
          <div
            class="pf-c-empty-state__icon"
          >
            <span
              aria-label="Contents"
              aria-valuetext="Loading..."
              class="pf-c-spinner pf-m-xl"
              role="progressbar"
            >
              <span
                class="pf-c-spinner__clipper"
              />
              <span
                class="pf-c-spinner__lead-ball"
              />
              <span
                class="pf-c-spinner__tail-ball"
              />
            </span>
          </div>
          <div
            class="pf-c-empty-state__body"
          >
            Loading ...
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`AccessConsoles with wrapped SerialConsole as a child 1`] = `
<DocumentFragment>
  <div
    class="pf-c-console"
  >
    <p>
      Serial console text
    </p>
  </div>
</DocumentFragment>
`;
